devices/src/virtio/base/ring.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: (Apache-2.0 AND BSD-3-Clause)
+
+//! Helper structs for working with VirtQueue rings.
+//!
+//! The virtqueue rings are dynamically sized types since they contain an array that has the
+//! size of the virtqueue, which can only be determined at runtime. For example this is the
+//! definition of the structure representing the virtqueue available ring:
+//!
+//! ```C
+//! struct virtq_avail {
+//!     le16 flags;
+//!     le16 idx;
+//!     le16 ring[ /* Queue Size */ ];
+//!     le16 used_event; /* Only if VIRTIO_F_EVENT_IDX */
+//! };
+//! ```
+//!
+//! Because of this constraint, we can't represent them with a simple statically-sized rust struct.
+//! However, we can leverage some of the properties of the rust structures in order to make it
+//! easier to work with these types.
+//!
+//! TODO: check if this can be implemented better after FixedSizeArray in CoerceUnsized
+//! become stable
+
+use memory_model::{GuestAddress, GuestMemory, GuestMemoryError};
+use std::mem::size_of;
+use std::ops::{Index, IndexMut};
+
+#[derive(Debug)]
+pub enum Error {
+    // An error generated by the GuestMemory module
+    GuestMemoryError(GuestMemoryError),
+    // the structure doesn't fit in the provided GuestMemory
+    InsufficientMemory,
+}
+
+/// Trait denoting that a rust struct is a Virtq ring mirror structure.
+///
+/// We can make it easier to work with virtqueue rings by defining mirror structures that respect
+/// the following properties:
+/// * they are PODs
+/// * they mirror all the members of the original structure found before the dynamically sized
+///   array. These members will be easily accessible as regular members of a structure.
+/// * instead of the dynamically sized array, they contain a zero sized slice of type `Element`.
+///   This affects the size of the mirror structure by including padding added to accommodate
+///   the array. This way we ensure that we can find the array at the end of the mirror structure.
+/// * `Element` is a POD.
+/// * they don't contain anything after the zero sized slice. All the members of the original
+///   structure found after the dynamically sized array can be handled by directly accessing
+///   the memory.
+///
+/// For example, a mirror structure for `virtq_avail` would look like this:
+///
+/// ```
+/// #[repr(C)]
+/// pub struct VirtqAvail {
+///     pub flags: u16,
+///     pub idx: u16,
+///     pub ring: [u16; 0],
+/// }
+/// ```
+///
+/// This trait is unsafe since there is no way to check that the implementor respects the
+/// properties listed above. Violating these constraints may cause problems.
+pub unsafe trait VirtqRingMirror {
+    // The type of the ring elements.
+    type Element;
+}
+
+#[derive(Debug)]
+pub struct VirtqRingPtr<T: VirtqRingMirror> {
```

Comment:
This struct, along with all others defined in this internal `mod`, don't need to be `pub`. `pub(crate)` is enough because they aren't - and shouldn't be - used outside the `devices` crate, and are not exported.